File: pml/sprites/registry.py
"""PML component registry — central registry for semantic components."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Callable

from pml.environment import Environment
from pml.graphics.objects import GraphicObject
from pml.types import BuiltinProcedure

if TYPE_CHECKING:
    from pml.sprites.validator import ParamSchema


class ComponentRegistry:
    """Registry for semantic sprite components (body, head, eyes, etc.)."""

    _instance: ComponentRegistry | None = None

    def __init__(self) -> None:
        self._components: dict[str, Callable[..., GraphicObject]] = {}
        self._schemas: dict[str, ParamSchema] = {}
        self._categories: dict[str, str] = {}

    @classmethod
    def get_instance(cls) -> ComponentRegistry:
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def register(
        self,
        name: str,
        factory: Callable[..., GraphicObject],
        schema: ParamSchema | None = None,
        category: str | None = None,
    ) -> None:
        self._components[name] = factory
        if schema is not None:
            self._schemas[name] = schema
        if category is not None:
            self._categories[name] = category

    def create(self, name: str, **kwargs: Any) -> GraphicObject:
        if name not in self._components:
            raise ValueError(f"Unknown component: {name}")
        return self._components[name](**kwargs)

    def has(self, name: str) -> bool:
        return name in self._components

    def list_components(self) -> list[str]:
        return list(self._components.keys())

    def get_schema(self, name: str) -> ParamSchema | None:
        return self._schemas.get(name)

    def get_category(self, name: str) -> str | None:
        return self._categories.get(name)

    def list_by_category(self, category: str) -> list[str]:
        return [n for n, c in self._categories.items() if c == category]

    def categories(self) -> list[str]:
        return sorted(set(self._categories.values()))


def register_components(env: Environment) -> None:
    """Register all semantic components as builtins."""
    # Phase 2+3: core character components
    from pml.sprites.components.body import create_body, _BODY_SCHEMA
    from pml.sprites.components.head import create_head, _HEAD_SCHEMA
    from pml.sprites.components.eyes import create_eyes, _EYES_SCHEMA
    from pml.sprites.components.mouth import create_mouth, _MOUTH_SCHEMA
    from pml.sprites.components.hair import create_hair, _HAIR_SCHEMA
    from pml.sprites.components.character import create_character, _CHARACTER_SCHEMA

    # Phase 4: extended components
    from pml.sprites.components.outfit import create_outfit, _OUTFIT_SCHEMA
    from pml.sprites.components.items import (
        create_weapon,
        create_potion,
        create_chest,
        create_generic_item,
        _WEAPON_SCHEMA,
        _POTION_SCHEMA,
        _CHEST_SCHEMA,
        _GENERIC_ITEM_SCHEMA,
    )
    from pml.sprites.components.ui_widgets import (
        create_button,
        create_panel,
        create_health_bar,
        create_icon,
        _BUTTON_SCHEMA,
        _PANEL_SCHEMA,
        _HEALTH_BAR_SCHEMA,
        _ICON_SCHEMA,
    )
    from pml.sprites.components.scene_elements import (
        create_tile,
        create_decoration,
        create_background,
        _TILE_SCHEMA,
        _DECORATION_SCHEMA,
        _BACKGROUND_SCHEMA,
    )

    registry = ComponentRegistry.get_instance()

    # (name, factory, schema, category)
    components = [
        # Character
        ("body", create_body, _BODY_SCHEMA, "character"),
        ("head", create_head, _HEAD_SCHEMA, "character"),
        ("anime-eyes", create_eyes, _EYES_SCHEMA, "character"),
        ("mouth", create_mouth, _MOUTH_SCHEMA, "character"),
        ("hair", create_hair, _HAIR_SCHEMA, "character"),
        ("character", create_character, _CHARACTER_SCHEMA, "character"),
        # Outfit
        ("outfit", create_outfit, _OUTFIT_SCHEMA, "character"),
        # Items
        ("weapon", create_weapon, _WEAPON_SCHEMA, "items"),
        ("potion", create_potion, _POTION_SCHEMA, "items"),
        ("chest", create_chest, _CHEST_SCHEMA, "items"),
        ("generic-item", create_generic_item, _GENERIC_ITEM_SCHEMA, "items"),
        # UI widgets
        ("button", create_button, _BUTTON_SCHEMA, "ui"),
        ("panel", create_panel, _PANEL_SCHEMA, "ui"),
        ("health-bar", create_health_bar, _HEALTH_BAR_SCHEMA, "ui"),
        ("icon", create_icon, _ICON_SCHEMA, "ui"),
        # Scene elements
        ("tile", create_tile, _TILE_SCHEMA, "scene"),
        ("decoration", create_decoration, _DECORATION_SCHEMA, "scene"),
        ("background", create_background, _BACKGROUND_SCHEMA, "scene"),
    ]

    for name, factory, schema, category in components:
        registry.register(name, factory, schema=schema, category=category)
        env.define(name, BuiltinProcedure(name, factory, accepts_kwargs=True))

File: src/pml/sprites/registry.cpp
// ==========================================================================================================================================================================================================================================═
// PML Component Registry — Implementation
// ==========================================================================================================================================================================================================================================═

#include "registry.h"

#include "error.h"
#include "objects.h"

#include "components/body.h"
#include "components/head.h"
#include "components/eyes.h"
#include "components/mouth.h"
#include "components/hair.h"
#include "components/character.h"
#include "components/outfit.h"
#include "components/items.h"
#include "components/ui_widgets.h"
#include "components/scene_elements.h"

#include <algorithm>
#include <format>
#include <set>

namespace pml {

// ==========================================================================================================================================================================================================================================═
// Singleton
// ==========================================================================================================================================================================================================================================═

ComponentRegistry& ComponentRegistry::instance() {
    static ComponentRegistry s_instance;
    return s_instance;
}

// ==========================================================================================================================================================================================================================================═
// Registration / lookup
// ==========================================================================================================================================================================================================================================═

void ComponentRegistry::register_(const std::string& name,
                                  ComponentFactory factory,
                                  std::optional<ParamSchema> schema,
                                  std::optional<std::string> category) {
    Entry entry;
    entry.factory = std::move(factory);
    entry.schema = std::move(schema);
    entry.category = std::move(category);
    m_components[name] = std::move(entry);
}

Value ComponentRegistry::create(
    const std::string& name,
    const std::unordered_map<std::string, Value>& kwargs) const {
    auto it = m_components.find(name);
    if (it == m_components.end()) {
        throw std::runtime_error(std::format("Unknown component: {}", name));
    }
    return it->second.factory(kwargs);
}

bool ComponentRegistry::has(const std::string& name) const {
    return m_components.find(name) != m_components.end();
}

std::vector<std::string> ComponentRegistry::list_components() const {
    std::vector<std::string> names;
    names.reserve(m_components.size());
    for (const auto& [name, _] : m_components) {
        names.push_back(name);
    }
    return names;
}

std::optional<ParamSchema> ComponentRegistry::get_schema(
    const std::string& name) const {
    auto it = m_components.find(name);
    if (it == m_components.end()) return std::nullopt;
    return it->second.schema;
}

std::optional<std::string> ComponentRegistry::get_category(
    const std::string& name) const {
    auto it = m_components.find(name);
    if (it == m_components.end()) return std::nullopt;
    return it->second.category;
}

std::vector<std::string> ComponentRegistry::list_by_category(
    const std::string& category) const {
    std::vector<std::string> names;
    for (const auto& [name, entry] : m_components) {
        if (entry.category == category) {
            names.push_back(name);
        }
    }
    return names;
}

std::vector<std::string> ComponentRegistry::categories() const {
    std::set<std::string> cats;
    for (const auto& [_, entry] : m_components) {
        if (entry.category) {
            cats.insert(*entry.category);
        }
    }
    return std::vector<std::string>(cats.begin(), cats.end());
}

// ==========================================================================================================================================================================================================================================═
// Component builtin registration
// ==========================================================================================================================================================================================================================================═

namespace {

/// Parse flat keyword arguments from a vector of Values.
[[nodiscard]] std::unordered_map<std::string, Value> parse_component_kwargs(
    const std::vector<Value>& args) {
    std::unordered_map<std::string, Value> result;
    for (size_t i = 0; i + 1 < args.size(); i += 2) {
        if (const auto* kw = args[i].as_keyword()) {
            result[kw->name] = args[i + 1];
        } else if (const auto* sym = args[i].as_symbol()) {
            result[sym->name] = args[i + 1];
        } else {
            break;
        }
    }
    return result;
}

/// Wrap a component factory into a BuiltinProcedure function.
[[nodiscard]] BuiltinProcedure::BuiltinFn wrap_factory(
    ComponentFactory factory) {
    return [factory = std::move(factory)](const std::vector<Value>& args,
                                          Environment&) -> Result<Value> {
        auto kwargs = parse_component_kwargs(args);
        return factory(kwargs);
    };
}

}  // anonymous namespace

void register_components(std::shared_ptr<Environment> env) {
    if (!env) return;

    auto& registry = ComponentRegistry::instance();

    // Helper to register a component both in the registry and as a builtin.
    auto reg = [&](const std::string& name,
                   ComponentFactory factory,
                   ParamSchema schema,
                   const std::string& category) {
        registry.register_(name, factory, schema, category);
        env->define(name,
            Value(std::make_shared<BuiltinProcedure>(
                name, wrap_factory(factory), true)));
    };

    // Character components
    reg("body", create_body, body_schema(), "character");
    reg("head", create_head, head_schema(), "character");
    reg("anime-eyes", create_eyes, eyes_schema(), "character");
    reg("mouth", create_mouth, mouth_schema(), "character");
    reg("hair", create_hair, hair_schema(), "character");
    reg("character", create_character, character_schema(), "character");
    reg("outfit", create_outfit, outfit_schema(), "character");

    // Items
    reg("weapon", create_weapon, weapon_schema(), "items");
    reg("potion", create_potion, potion_schema(), "items");
    reg("chest", create_chest, chest_schema(), "items");
    reg("generic-item", create_generic_item, generic_item_schema(), "items");

    // UI widgets
    reg("button", create_button, button_schema(), "ui");
    reg("panel", create_panel, panel_schema(), "ui");
    reg("health-bar", create_health_bar, health_bar_schema(), "ui");
    reg("icon", create_icon, icon_schema(), "ui");

    // Scene elements
    reg("tile", create_tile, tile_schema(), "scene");
    reg("decoration", create_decoration, decoration_schema(), "scene");
    reg("background", create_background, background_schema(), "scene");
}

}  // namespace pml
